compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/CoroutineCodegen.kt

```diff
@@ -154,62 +147,43 @@ private fun IrFunction.suspendFunctionView(context: JvmBackendContext): IrFuncti
     // And we need to generate this function for callable references
     val isBridgeInvokeOfCallableReference = origin == IrDeclarationOrigin.BRIDGE &&
             (parent as? IrClass)?.origin == JvmLoweredDeclarationOrigin.FUNCTION_REFERENCE_IMPL
-    val descriptor =
-        if (originalDescriptor is DescriptorWithContainerSource && originalDescriptor.containerSource != null)
-            WrappedFunctionDescriptorWithContainerSource(originalDescriptor.containerSource!!)
-        else
-            WrappedSimpleFunctionDescriptor(sourceElement = originalDescriptor.source)
-    return IrFunctionImpl(
-        startOffset, endOffset,
-        if (origin == JvmLoweredDeclarationOrigin.FOR_INLINE_STATE_MACHINE_TEMPLATE_CAPTURES_CROSSINLINE)
+    val continuationType = if (isInvokeOfNumberedSuspendFunction || isBridgeInvokeOfCallableReference)
+        context.irBuiltIns.anyNType
+    else
+        context.ir.symbols.continuationClass.createType(false, listOf(makeTypeProjection(returnType, Variance.INVARIANT)))
+    return buildFunWithDescriptorForInlining(originalDescriptor) {
+        updateFrom(this@suspendFunctionView)
+        name = this@suspendFunctionView.name
+        origin = if (origin == JvmLoweredDeclarationOrigin.FOR_INLINE_STATE_MACHINE_TEMPLATE_CAPTURES_CROSSINLINE)
             JvmLoweredDeclarationOrigin.FOR_INLINE_STATE_MACHINE_TEMPLATE_CAPTURES_CROSSINLINE_VIEW
         else
-            JvmLoweredDeclarationOrigin.SUSPEND_FUNCTION_VIEW,
-        IrSimpleFunctionSymbolImpl(descriptor),
-        name, visibility, modality, context.irBuiltIns.anyNType,
-        isInline = isInline, isExternal = isExternal, isTailrec = isTailrec, isSuspend = isSuspend, isExpect = isExpect,
-        isFakeOverride = false, isOperator = false
-    ).also {
-        descriptor.bind(it)
+            JvmLoweredDeclarationOrigin.SUSPEND_FUNCTION_VIEW
+        returnType = context.irBuiltIns.anyNType
+    }.also {
         it.parent = parent
 
         it.annotations.addAll(annotations)
 
+        it.copyAttributes(this)
         it.copyTypeParametersFrom(this)
-
-        it.dispatchReceiverParameter = dispatchReceiverParameter?.copyTo(it)
-        it.extensionReceiverParameter = extensionReceiverParameter?.copyTo(it)
-
-        valueParameters.mapTo(it.valueParameters) { p -> p.copyTo(it) }
-        it.addValueParameter(
-            SUSPEND_FUNCTION_COMPLETION_PARAMETER_NAME,
-            if (isInvokeOfNumberedSuspendFunction || isBridgeInvokeOfCallableReference) context.irBuiltIns.anyNType
-            else context.ir.symbols.continuationClass.createType(false, listOf(makeTypeProjection(returnType, Variance.INVARIANT)))
-        )
-        val valueParametersMapping = explicitParameters.zip(it.explicitParameters).toMap()
+        it.copyValueParametersFrom(this)
+        // TODO: this should come before the default masks, if they exist; otherwise, this is not binary
+        //       compatible with the non-IR backend
```

Comment:
@udalov IMO it's probably time to think about those backwards compatibility tests